Add pump model and correct RC for volume forces

tanoret opened the pull request:
closes #24775 #24776

tanoret committed:
Add pump model and correct RC for volume forces Refs #24775 #24776

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1605219/) on 8d54d2a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dd3ae2a0b7da2bee80f465761e44ba82300830b4`

tanoret commented:
Also, here are the new solutions with corrected RC for issue #24776 

<img width="658" alt="image" src="https://github.com/idaholab/moose/assets/29891082/ac82a5e4-00b4-4377-b93c-0c80e42a3092">

<img width="658" alt="image" src="https://github.com/idaholab/moose/assets/29891082/652a13e1-629d-4152-b1b1-6ce345c7214c">

GiudGiud commented on modules/navier_stokes/src/materials/NSFVPump.C:
```diff
@@ -0,0 +1,136 @@
+
+#include "NSFVPump.h"
+#include "NS.h"
+
```
```suggestion
#include "Function.h"

```

tanoret replied:
Thanks!

tanoret commented:
Also, here is the solution for #24776 with the corrected RC coefficients

<img width="658" alt="image" src="https://github.com/idaholab/moose/assets/29891082/30b15665-67f6-4e89-ba31-242c25f42249">

<img width="658" alt="image" src="https://github.com/idaholab/moose/assets/29891082/168593db-6a8b-4754-bf2f-5693fbf570e8">

tanoret committed:
adding function to NSFVPump Refs #24775 #24776

GiudGiud commented:
These results look awesome! Does it influence the head of the pump to turn on/off the correction?

tanoret commented:
> These results look awesome! Does it influence the head of the pump to turn on/off the correction?

The RC correction does not influence too much the head of the pump. Even oscillating, the total pressure drop for the standard RC coefficients is quite right

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1606186/) on df1547a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d2a66a</th>
      <th colspan=4>#24778 df1547</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.41%</td>
      <td>85.41%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88823</td>
      <td>88827</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15178</td>
      <td>15174</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24778/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24778/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>d2a66a</th>
      <th colspan=4>#24778 df1547</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.85%</td>
      <td>83.97%</td>
      <td>+0.11%</td>
      <td>91.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10827</td>
      <td>11002</td>
      <td>+175</td>
      <td>176</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2085</td>
      <td>2101</td>
      <td>+16</td>
      <td>16</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24778/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24778/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24778/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1606145/) on df1547a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24778/site)

This comment will be updated on new commits.

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/pump/tests:
```diff
@@ -0,0 +1,49 @@
+    type = 'CSVDiff'
+    input = 'pump_and_counterpump_loop.i'
+    csvdiff = 'pump_and_counterpump_loop_out.csv'
+    requirement = 'The system shall mantain the correct flowand maximum speed (no oscillations) with two competing pumps.'
```
```suggestion
    requirement = 'The system shall maintain the correct flow and maximum speed (no oscillations) with two competing pumps.'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/pump/tests:
```diff
@@ -0,0 +1,49 @@
+    type = 'CSVDiff'
+    input = 'pump_loop_negative_rotation.i'
+    csvdiff = 'pump_loop_negative_rotation_out.csv'
+    requirement = 'The system shall mantain the correct flow and maximum speed (no oscillations) with one backwards rotating pump.'
```
```suggestion
    requirement = 'The system shall maintain the correct flow and maximum speed (no oscillations) with one backwards rotating pump.'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/pump/tests:
```diff
@@ -0,0 +1,49 @@
+    type = 'CSVDiff'
+    input = 'pump_loop.i'
+    csvdiff = 'pump_loop_out.csv'
+    requirement = 'The system shall mantain the correct flow and maximum speed (no oscillations) with one forward rotating pump.'
```
```suggestion
    requirement = 'The system shall maintain the correct flow and maximum speed (no oscillations) with one forward rotating pump.'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/pump/tests:
```diff
@@ -0,0 +1,49 @@
+    type = 'CSVDiff'
+    input = 'pump_as_volume_force_loop_pressure_corrected.i'
+    csvdiff = 'pump_as_volume_force_loop_pressure_corrected_out.csv'
+    requirement = 'The system shall mantain the correct flow and maximum speed (no oscillations) with pressure-correction to Rhie-Chow.'
```
```suggestion
    requirement = 'The system shall maintain the correct flow and maximum speed (no oscillations) with a body force correction to Rhie-Chow computed by examining the difference between two different kinds of body force interpolation to faces.'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/pump/tests:
```diff
@@ -0,0 +1,49 @@
+    type = 'CSVDiff'
+    input = 'pump_as_volume_force_loop_force_corrected.i'
+    csvdiff = 'pump_as_volume_force_loop_force_corrected_out.csv'
+    requirement = 'The system shall mantain the correct flow and maximum speed (no oscillations) with force-correction to Rhie-Chow.'
```
```suggestion
    requirement = 'The system shall maintain the correct flow and maximum speed (no oscillations) with a body force correction to Rhie-Chow computed by checking for a non-zero body force face interpolation.'
```

lindsayad reviewed (changes requested):
really awesome contribution

lindsayad commented:
This PR adds important capability and should remain open

tanoret committed:
Address some review points

- Typo/Grammar fixes
- Make more data members const
- Make some requirements more descriptive
- Remove unnecessary new line

Refs #24775 #24776

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

tanoret commented on modules/navier_stokes/include/materials/NSFVPump.h:
```diff
@@ -0,0 +1,67 @@
+  /// The pressure head should be provided as head (in meters) vs. flow rate (in m3/s)
+  const Function * _pressure_head_function;
+
+  /// Rated transversal area of the pump
+  const Moose::Functor<Real> & _area_rated;
+
+  /// Rated volume of the pump
+  const Moose::Functor<Real> & _volume_rated;
```
They could be in case we want to introduce a space function or a volume-smoothed area change in the pump. I find it more general this way. What do you think?

lindsayad replied:
works for me

tanoret committed:
Rework RC UO logic

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

tanoret committed:
Make more local variables const

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

tanoret commented on modules/navier_stokes/src/materials/NSFVPump.C:
```diff
@@ -0,0 +1,137 @@
+            rated_pressure_head * std::pow(flow_rate_scaling * rotation_speed_scaling, 4.0 / 3.0);
+
+        // Computing effective volume force
+        const auto rho = _rho(r, t);
+        const Real gravity = _gravity.norm();
+        const auto area = _area_rated(r, t);
+        const auto volume = _volume_rated(r, t);
+        return raw_value(-rho * gravity * pressure_head * area / volume);
```
Thanks. Added the suggested warning

tanoret committed:
Assert density is constant so we can remove `raw_value` call

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

tanoret committed:
Catch sets as references

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1813511/) on 1897a60 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ab89aee763cddf8689693629d508ad63a87bcd79`

tanoret committed:
Remove `raw_value(rho)` from pump force calculation

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

tanoret committed:
More descriptive parameter docstring

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

tanoret committed:
averaging_method -> volume_force_correction_method (or similar)

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

tanoret committed:
RC UO code refactoring

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1813854/) on 92d0d19 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ab89aee763cddf8689693629d508ad63a87bcd79`

Commit:
Renaming NSFVPumpMaterial Refs #24775 #24776

tanoret commented on modules/navier_stokes/doc/content/source/materials/NSFVPump.md:
Renamed. Thanks

tanoret commented on modules/navier_stokes/include/materials/NSFVPump.h:
Renamed. Thanks.

tanoret commented on modules/navier_stokes/src/materials/NSFVPump.C:
Renamed. Thanks!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1814272/) on f0f8b41 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ab89aee763cddf8689693629d508ad63a87bcd79`

tanoret commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -601,6 +645,170 @@ INSFVRhieChowInterpolator::getVelocity(const Moose::FV::InterpMethod m,
+      for (const auto i : make_range(_volumetric_force.size()))
+        value += (*_volumetric_force[i])(loc_face, time) * (loc_normal * fi.normal());
+    }
+    if ((std::abs(value) - _baseline_volume_force) > 0)
```
The issue is that if for some reason the force is constant and non-zero everywhere, we will deactivate the RC coefficients for all domain and go back into checkerboarding

lindsayad replied:
In your comment above:

> In thery, pressure and velocity cannot be decoupled when a body force is present

it sounded like we don't need RC if there is any kind of body force which is why I asked. It sounds like that comment is not quite right?

tanoret replied:
Yes, you are right. Will delete that comment. Thanks

tanoret commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -601,6 +645,170 @@ INSFVRhieChowInterpolator::getVelocity(const Moose::FV::InterpMethod m,
+
+        for (const auto i : make_range(_volumetric_force.size()))
+          elem_value += (*this->_volumetric_force[i])(loc_face, time) * fi_loc->faceArea() *
+                        (fi_loc->faceCentroid() - elem->vertex_average()).norm() *
```
This is because I am trying to reconstruct the effective volume term from the faces. The volume associated to the face should be faceArea*||face_position - element_position||. Changed the line to `fi_loc->faceCentroid() - elem->centroid()).norm()` to be clear

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -601,6 +645,170 @@ INSFVRhieChowInterpolator::getVelocity(const Moose::FV::InterpMethod m,
+    Real uncorrected_interp_vf;
+
+    // Compute the corrected interpolated face value
+    if (Moose::FV::onBoundary(*this, fi))
+    {
+      const Elem * const boundary_elem = hasBlocks(elem->subdomain_id()) ? elem : neighbor;
+      const Moose::FaceArg boundary_face{
+          &fi, Moose::FV::LimiterType::CentralDifference, true, correct_skewness, boundary_elem};
+
+      interp_vf = 0.0;
+      for (const auto i : make_range(_volumetric_force.size()))
+        interp_vf += (*this->_volumetric_force[i])(boundary_face, time);
+    }
```
On what is line 650 now, we return when `Moose::FV::onBoundary` so we should never get into these lines

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -639,6 +696,166 @@ INSFVRhieChowInterpolator::getVelocity(const Moose::FV::InterpMethod m,
+          _moose_mesh.faceInfo(elem_has_fi ? elem : loc_neighbor,
+                               elem_has_fi ? side : loc_neighbor->which_neighbor_am_i(elem));
+
+      if (Moose::FV::onBoundary(*this, *fi_loc))
+      {
+        const Elem * const boundary_elem = hasBlocks(elem->subdomain_id()) ? elem : loc_neighbor;
+        const Moose::FaceArg loc_boundary_face{fi_loc,
+                                               Moose::FV::LimiterType::CentralDifference,
+                                               true,
+                                               correct_skewness,
+                                               boundary_elem};
+
+        for (const auto i : make_range(_volumetric_force.size()))
+          elem_value += (*this->_volumetric_force[i])(loc_boundary_face, time) *
+                        fi_loc->faceArea() *
+                        (fi_loc->faceCentroid() - elem->vertex_average()).norm() *
+                        (fi_loc->normal() * unit_basis_vector);
+      }
```
As above we should never hit these lines of code

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -639,6 +696,166 @@ INSFVRhieChowInterpolator::getVelocity(const Moose::FV::InterpMethod m,
+          _moose_mesh.faceInfo(elem_has_fi ? neighbor : loc_elem,
+                               elem_has_fi ? side : loc_elem->which_neighbor_am_i(neighbor));
+
+      if (Moose::FV::onBoundary(*this, *fi_loc))
+      {
+        const Elem * const boundary_elem =
+            hasBlocks(neighbor->subdomain_id()) ? neighbor : loc_elem;
+        const Moose::FaceArg loc_boundary_face{fi_loc,
+                                               Moose::FV::LimiterType::CentralDifference,
+                                               true,
+                                               correct_skewness,
+                                               boundary_elem};
+
+        for (const auto i : make_range(_volumetric_force.size()))
+          neigh_value += (*this->_volumetric_force[i])(loc_boundary_face, time) *
+                         fi_loc->faceArea() *
+                         (fi_loc->faceCentroid() - neighbor->vertex_average()).norm() *
+                         (fi_loc->normal() * unit_basis_vector);
+      }
```
should never hit these lines either

tanoret commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -601,6 +645,170 @@ INSFVRhieChowInterpolator::getVelocity(const Moose::FV::InterpMethod m,
+                        (fi_loc->normal() * loc_normal);
+      }
+    }
+    elem_value /= elem->volume();
```
Thanks added.

tanoret commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -601,6 +645,170 @@ INSFVRhieChowInterpolator::getVelocity(const Moose::FV::InterpMethod m,
+    // to the face
+    interpolate(
+        Moose::FV::InterpMethod::Average, uncorrected_interp_vf, elem_value, neigh_value, fi, true);
+    uncorrected_interp_vf *= fi.gC();
```
Yes. It is to make the two formulations consistent. I tried taking it out, but this yield back the pump oscillations

lindsayad replied:
It feels asymmetric. If one element centroid is closer to the face than the other, then the result of this calculation would change depending on the element numbering of the mesh (since we associate gC() with the lower ID element).

lindsayad replied:
It looks like this multiplication got removed which is cool. Are things still working as desired?

tanoret replied:
Thanks. This line was removed.

tanoret commented on modules/navier_stokes/test/tests/finite_volume/ins/pump/pump_and_counterpump_loop.i:
```diff
@@ -0,0 +1,267 @@
+[]
+
+[Problem]
+  material_coverage_check = False
```
Yes. It is because I don't want to define a material for the blocks that are not pumps

tanoret committed:
address face interpolation Refs #24775 #24776

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1824749/) on 8b1b757 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6d70fba25568259d2b64e247e42717a33ce7778a`

tanoret committed:
clang style changes Refs #24775 #24776

lindsayad commented:
@tanoret this needs to be rebased and have the merge conflict fixed. It would also be good to squash some of the “suggested changes” commits. I can handle that if you don’t want to

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1834173/) on 34d320f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24778/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3d2af038c621e2e6715218cb94fdc20d7596294f`

lindsayad commented on modules/navier_stokes/doc/content/source/fvkernels/INSFVPump.md:
```diff
@@ -0,0 +1,10 @@
+# INSFVPump
+
+This object sets the volumetric force of the pump in the momentum system
+via the pump functor provided by [NSFVPump.md].
```
```suggestion
via the pump functor provided by [NSFVPumpMaterial.md].
```
This should fix the failing documentation

lindsayad reviewed (changes requested):
would still be nice to squash these single-line commits into a more informative commit

<img width="1199" alt="Screenshot 2023-10-21 at 2 31 16 PM" src="https://github.com/idaholab/moose/assets/10248304/204f5fd0-e364-4407-9844-732ea4278583">

lindsayad committed:
NSFVPump -> NSFVPumpMaterial

lindsayad committed:
Add new params to RC unique params

This allows more easy incorporation of these params into the
action

lindsayad committed:
Add back mesh velocity
